### ISSUE TITLE
[RFC] On regular expression lookarounds

### ISSUE THREAD

**Issue by jdkato** [vale-cli/vale#233]
This is one of the most common feature requests / misunderstandings about how Vale's rules work. So, I'd like to take some time to try to get this right&mdash;whether that means changing existing functionality or simply improving the documentation on the topic.

## Why things are the way they are

To provide some background: Vale is written in [Go](https://golang.org/) and uses its stdlib [`regexp`](https://golang.org/pkg/regexp/syntax/) package to evaluate all rules. The implementation is based on Google's [RE2](https://github.com/google/re2) library:

> RE2 is a fast, safe, thread-friendly alternative to backtracking regular expression engines like those used in PCRE, Perl, and Python. It is a C++ library.

If you'd like to learn more about the decisions that went into RE2, the [wiki on `google/re2`](https://github.com/google/re2/wiki/WhyRE2) goes into some detail.

## A different take, but not a limitation

The academics of the issue aside, my stance has always been that I don't want this to be a limitation of Vale; all else equal, I'd rather Vale be a little slower but more capable.

I've written *a lot* of general-purpose regex (using many languages) and even more Vale-specific regex, and I've yet to find something that seemed impossible or unnecessarily difficult to implement with Vale in comparison to other tools. This leads me to believe that the problem is more related to documentation than implementation.

The [documentation](https://docs.errata.ai/vale/styles#extension-points) currently mentions the limitation of *Go* and links to a blog post that goes into detail about how Vale approaches the problem. But it's clear that this could be explained better.

## Go's stdlib regex engine doesn't support lookarounds, *but Vale does*

Vale supports both *positive* and *negative* lookarounds. It does so by taking advantage of the YAML wrapper around its rules.

### Positive lookarounds

The first variation of lookarounds are the *positive* lookahead and lookbehind assertions:

- lookahead (`x(?=y)`): "I want to match `x` only if it *precedes* `y` without matching `y`."
- lookbehind (`(?<=y)x`): "I want to match `x` only if it is *preceded* by `y` without matching `y`."

For Vale’s purposes, this should be straightforward to implement: 

```yaml
extends: existence
message: "Consider removing 'dialog'."
level: warning
ignorecase: true
tokens:
  - 'dialog (?:box)'
```

The basic idea here is that you don't have show the end-user every thing that the rule matched; you can use the `message` to gain control over what you show.

### Negative lookarounds

The second variation of lookarounds are the *negative* lookahead and lookbehind assertions:

- lookahead (`x(?!y)`): "I want to match `x` only if it does *not precede* `y`.”
- lookbehind (`(?<!y)x`): "I want to match `x` only if it is *not preceded* by `y`*.*"

For Vale’s purposes, these cases can be emulated using `substitution` checks:

```yaml
extends: substitution
message: "Use '%s' instead of '%s'"
level: error
ignorecase: false
swap:
  '(?:[^\s]*) ?VS Code': Microsoft VS Code
```

This rule will flag any instance of "VS Code" that isn't preceded by "Microsoft"; it's functionally equivalent to `(?<!Micrsoft )VS Code`. The key takeaway is that you can write patterns that *include* the accepted case since Vale will only raise alerts for those that don't match it.

So, instead of looking for `(?<!Micrsoft )` (anything that isn't 'Microsoft'), you can look for *anything* and Vale will compare it to the user-provided accepted case.

## Feedback

I know I haven't written rules that cover every possible use case, so I'd be quite interested in seeing examples of rules you either can't implement or find hard to implement without using raw lookarounds.

@ravlen: I know this specifically came up in your research session. Do you have any thoughts on the existing support or hard-to-implement examples you could share?

**Comment by Ravlen** [comment]
@jdkato I understand the workarounds (and actually saw them in the docs originally), but my desire to have workarounds is mainly because I instinctively try to use them when working with regex, and it's always frustrating then to realize "Oh yeah, I can't do it this way with Vale, let me go see how that workaround works".

Of course, probably the biggest part is that (as you know) we are already stretching Vale beyond your standard use cases, by making liberal use of `scope: raw`. Often the lookarounds I want to use are related to formatting, which is a bigger problem for us right now. If I'm using `scope: raw`, I don't think I can use the workarounds, right?

What I've had to do is write rules that look for "known incorrect usage". For example, if we have an alert box, it should be styled exactly as the following `NOTE: **Note:**`. What our rule does is look for `NOTE: **NOTE:**` and `NOTE: **note:**` (incorrect capitalization). But this does not catch cases where the wrong word is used, like `NOTE: **Tip:**`. It would be much better and more powerful to have negative lookahead that says basically `If NOTE: is not followed by **Note:** then raise a warning`, which looks like `NOTE:(?! \*\*Note:\*\*)`. I can't figure out how to do that with the regex we have access to now.

Still, as I've mentioned before, we *love* the `scope: raw` system, and it was a game changer when you added it. We've since added about 10 rules that let us remove literally thousands of incorrect formatting/linking/etc and lock them down so more aren't introduced. If we could somehow turn on lookarounds, like with a flag, `vale --lookarounds *.md`, that would be *huge*.

I'll ask Amy if she remembers any of the times we wanted lookarounds, but in regular text (not `scope: raw` examples), because I know we had some a few months ago.

**Comment by jdkato** [comment]
> I understand the workarounds (and actually saw them in the docs originally), but my desire to have workarounds is mainly because I instinctively try to use them when working with regex, and it's always frustrating then to realize "Oh yeah, I can't do it this way with Vale, let me go see how that workaround works".

This is somewhat of a semantic nitpick, but I don't think "workaround" is the correct word here. It implies that non-backtracking engines are somehow inadequate when, in reality, I think they simply take a *different* approach (which is becoming increasingly more common: RE2, Go, Rust, etc.).

> If I'm using `scope: raw`, I don't think I can use the workarounds, right?

The only features that you can't use with `scope: raw` are those that require HTML-related processing (which `raw` disables). All regex features will work with any scope.

> What I've had to do is write rules that look for "known incorrect usage". For example, if we have an alert box, it should be styled exactly as the following `NOTE: **Note:**`. What our rule does is look for `NOTE: **NOTE:**` and `NOTE: **note:**` (incorrect capitalization). But this does not catch cases where the wrong word is used, like `NOTE: **Tip:**`. It would be much better and more powerful to have negative lookahead that says basically If `NOTE: is not followed by **Note:** then raise a warning`, which looks like `NOTE:(?! \*\*Note:\*\*)`. I can't figure out how to do that with the regex we have access to now.

This is a great example of when to use `substitution`:

```yaml
extends: substitution
message: "Use '%s' instead of '%s'."
level: error
# We don't want word boundaries (`\b`) here.
nonword: true
scope: raw
swap:
  'NOTE: \*\*\w+:\*\*': 'NOTE: **Note:**'
```
This is *equivalent* to `NOTE: (?!\*\*Note:\*\*)`: the key takeaway is that the left-hand pattern *can match* the right-hand value. Vale won't raise an alert if the pattern finds the expected value, which is something that you can't do with regex alone (backtracking or otherwise). 

In other words, since Vale has access to more information than regex alone would, `?!` is really no longer needed. Here's an example of the rule in action:

![Screen Shot 2020-09-16 at 12 08 59 PM](https://user-images.githubusercontent.com/8785025/93381478-940b6d80-f815-11ea-9f4e-981422946bdd.png)

In general, when you feel you need a negative lookaround, all you need to rememeber is: (1) use `substitution` and (2) Vale doesn't raise alerts for patterns that match the expected case.

**Comment by Ravlen** [comment]
@jdkato Ahh, ok, I see how that's working now, that will certainly be useful now that I understand it a bit better! That point: `the key takeaway is that the left-hand pattern can match the right-hand value. Vale won't raise an alert if the pattern finds the expected value` is what I was missing. 🙇 I will try to apply that to some of our rules! 👍 

A followup question: how would you do the same thing for when more than one result is acceptable? I didn't mention this in my first post, because I thought I shouldn't get into the weeds too much... I might as well ask though, because again I can't imaging how to do it.

If you check our style guide, `NOTE:` and `TIP:` only have one acceptable word, so the technique above works for those (and I'll make use of them): https://docs.gitlab.com/ee/development/documentation/styleguide.html#alert-boxes

 On the other hand `CAUTION:` and `DANGER:` allow a set of words (but no more than those words). In this case, my instinct would be to use same regex as I thought before (`NOTE: (?!\*\*Note:\*\*)`), but add capture groups for all the acceptable words:

- `CAUTION: (?!\*\*(Caution|Warning|Important):\*\*)`

In the docs though, I do see that nested capture groups aren't possible: https://docs.errata.ai/vale/styles#substitution

Is there a _different approach_ (😉 ) for Vale in this case?

I'm not seeing it right away, but would it make sense to ask for a feature where substitution allows multiple swap values? Like:

```yaml
extends: substitution
message: "The alert box text for CAUTION: can only use Caution:, Warning:, or Important:."
level: error
# We don't want word boundaries (`\b`) here.
nonword: true
scope: raw
swap:
  'CAUTION: \*\*\w+:\*\*': 'CAUTION: **Caution:**', 'CAUTION: **Warning:**', 'CAUTION: **Important:**'
```

I'm very busy this week, so I can't dig too much into our current rules right now, but next week I'll see how much I can take care of with the current feature set, and I'll see if I can find other examples that might still be falling through the cracks.

**Comment by jdkato** [comment]
> On the other hand `CAUTION:` and `DANGER:` allow a set of words (but no more than those words). In this case, my instinct would be to use same regex as I thought before `(NOTE: (?!\*\*Note:\*\*))`, but add capture groups for all the acceptable words:
>
> - `CAUTION: (?!\*\*(Caution|Warning|Important):\*\*)`
>
> In the docs though, I do see that nested capture groups aren't possible: https://docs.errata.ai/vale/styles#substitution
>
> Is there a different approach (😉 ) for Vale in this case?

I think this is a general regex jargon issue. Your example pattern doesn't need a *capture* group (of which it only has one currently); it needs alternation (the use of `|`).

The pattern could be written more optimally (in terms of performance) as `CAUTION: (?!\*\*(?:Caution|Warning|Important):\*\*)`, in which case it uses a *non-capturing* group (which Vale supports).

A good "rule of thumb," if you will, when writing regex for any engine is to always use a *non-capturing* group (`(?: ... )`) unless you specifically need to make reference to a captured element (either later in the pattern or in your application logic).

--- 

For Vale's purposes, I think the ideal rule would be:

```yaml
extends: substitution
message: "The alert box text for CAUTION: can only use 'Caution:', 'Warning:', or 'Important:'."
level: error
nonword: true
scope: raw
swap:
  'CAUTION: \*\*\w+:\*\*': 'CAUTION: \*\*(?:Caution|Warning|Important):\*\*'
```

This currently isn't possible because the right-hand side only performs simple string comparison, but I've added support for this in https://github.com/errata-ai/vale/commit/4b5a384b2fe245500e3c1eeb379f3e9a0e9b4ebb. 

So, this will be possible in v2.4.0. I'll also work on making the docs on regular expressions more clear.

**Comment by Ravlen** [comment]
@jdkato This is all awesome stuff, thanks a lot, I will see if I can dedicate a day to trying to fix all our rules! https://github.com/errata-ai/vale/commit/4b5a384b2fe245500e3c1eeb379f3e9a0e9b4ebb looks especially interesting, I really like the potential of that!

Thanks a gain, this is brilliant!

Edit: Not just the new feature, but also the explanation of how to use Vale *properly*. 🙇

**Comment by nostodnayr** [comment]
I'm having trouble finding a way to implement a rule that disallows 'data' on its own but allows 'data breach'. In my natural habitat, I'd write `data(?! breach)`, but I'm perhaps too thick to understand how to do this with the tools at hand in Go/Vale.

**Comment by jdkato** [comment]
@nostodnayr:

```yaml
extends: substitution
message: "Don't use 'data'."
level: error
swap:
  'data ?(?:[^\s]*)': 'data breach'
```

As mentioned above, the key takeaway is that the left-hand pattern can match the right-hand value (in this case, 'data breach') without raising unwanted errors.

**Comment by nostodnayr** [comment]
@jdkato ah, sorry I didn’t spell that out fully. I want to keep a substitution of ‘data’ to ‘information’, but not substitute anything for ‘data breach’. 

Edit: thank you, regardless for your saint-like patience with me. :-)
Edit 2: In this case, additionally, words like 'database' also get caught in the rule as written.
Edit 3: I think I've solved this, using the caution box example above as a model. This works in my limited testing. Any thoughts? 

    level      : warning
    message    : Rewrite to avoid "data", or use information if it makes sense in the context. 

    extends: substitution
    nonword: true
    scope: raw

    swap:
      'data\b ?\w+': 'data (?:breach)'

**Comment by jdkato** [comment]
I would write it like this:

```yaml
extends: substitution
message: 'Rewrite to avoid "data", or use "information" if it makes sense in the context.'
level: warning
ignorecase: true
swap:
  'data ?(?:[^\s]*)': 'data breach'
```
Most notably, you don't need `scope: raw`.

**Comment by jdkato** [comment]
After hearing from a number of users (here, on Slack, and through #220), I've decided to pursue bringing PCRE-compatible regex to Vale (lookarounds and backreferences, most notably).

I've outlined my plans in more detail in the post [*The future of Vale: Announcing the 3x10 funding campaign*](https://jdkato.medium.com/the-future-of-vale-announcing-the-3x10-funding-campaign-5e665bac974a), which explains what I'd like to do and announces a fundraising campaign to allow me to dedicate more time to making significant changes to Vale.

**Comment by jdkato** [comment]
As of [v2.9.0](https://github.com/errata-ai/vale/releases/tag/v2.9.0), "PCRE-style" patterns are supported in `existence`. Support in other extension points will be incrementally added.

**Comment by Ravlen** [comment]
This. Is. Huge! Thanks so much, @jdkato, you are a miracle worker!

**Comment by rainecheck** [comment]
Hi, quick question related to the above - does the support for PCRE-style patterns extend to use in the `accept.txt` and `reject.txt` files, or just the `.yaml` files?

**Comment by jdkato** [comment]
@rainecheck, you should be able to use the same regex anywhere.
